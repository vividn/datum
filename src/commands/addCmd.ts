import { EitherDocument } from "../documentControl/DatumDocument";
import { connectDb } from "../auth/connectDb";
import {
  addDoc,
  conflictChoices,
  ConflictStrategyNames,
} from "../documentControl/addDoc";
import { dataArgs, DataArgs, handleDataArgs } from "../input/dataArgs";
import { MainDatumArgs } from "../input/mainArgs";
import { addIdAndMetadata } from "../meta/addIdAndMetadata";
import { primitiveUndo } from "../undo/primitiveUndo";
import { FieldArgs, fieldArgs } from "../input/fieldArgs";
import { flexiblePositional } from "../input/flexiblePositional";
import { updateLastDocsRef } from "../documentControl/lastDocs";
import { compileState } from "../state/compileState";
import { ArgumentParser } from "argparse";
import { dbArgs } from "../input/dbArgs";
import { outputArgs } from "../input/outputArgs";
import { parseIfNeeded } from "../utils/parseIfNeeded";

export const newDocArgs = new ArgumentParser({
  add_help: false,
});
newDocArgs.add_argument("--no-metadata", "-M", {
  help: "do not include meta data in document",
  action: "store_true",
  dest: "noMetadata",
});
newDocArgs.add_argument("--id-part", "--id", {
  help:
    "Which field(s) to use for the _id field in the document." +
    " Can either be a single string with fields delimited by --id-delimiter" +
    " or can be used multiple times to progressively assemble an id delimited by --id-delimiter",
  action: "append",
  dest: "idParts",
});
newDocArgs.add_argument("--id-delimiter", {
  help: "spacer between fields in the id",
});
newDocArgs.add_argument("--undo", "-u", {
  help: "undoes the last datum entry",
  action: "store_true",
});
newDocArgs.add_argument("--force-undo", "-U", {
  help: "forces an undo, even if the datapoint was entered more than 15 minutes ago",
  action: "store_true",
  dest: "forceUndo",
});
newDocArgs.add_argument("--merge", "-x", {
  help: "on conflict with an existing document update with the merge strategy. Equivalent to `--update merge`",
  action: "store_const",
  const: "merge",
  dest: "conflict",
});
newDocArgs.add_argument("--conflict", "-X", {
  help: `on conflict, update with given strategy.`,
  choices: conflictChoices,
});

export const addArgs = new ArgumentParser({
  add_help: false,
  parents: [fieldArgs, newDocArgs, dataArgs],
});
export const addCmdArgs = new ArgumentParser({
  description: "add a document",
  prog: "datum add",
  usage: `%(prog)s <field> [data..]
  %(prog)s --fieldless [data..]
  %(prog)s <field> -k <reqKey> -k <optKey>=[defaultValue] ... <reqValue> [optValue] [data..]
`,
  parents: [addArgs, outputArgs, dbArgs],
});

export type AddCmdArgs = MainDatumArgs &
  FieldArgs &
  DataArgs & {
    noMetadata?: boolean;
    idParts?: string[];
    idDelimiter?: string;
    undo?: boolean;
    forceUndo?: boolean;
    merge?: boolean;
    conflict?: ConflictStrategyNames;
  };

export async function addCmd(
  args: AddCmdArgs | string | string[],
  preparsed?: Partial<AddCmdArgs>,
): Promise<EitherDocument> {
  args = parseIfNeeded(addCmdArgs, args, preparsed);
  const db = connectDb(args);

  flexiblePositional(args, "field", "field", args.fieldless);

  const payloadData = handleDataArgs(args);

  const payloadWithState = await compileState(db, payloadData);

  const payload = addIdAndMetadata(payloadWithState, args);

  // update now in case the addDoc fails due to conflict
  await updateLastDocsRef(db, payload._id);

  const { undo, forceUndo } = args;
  if (undo || forceUndo) {
    return await primitiveUndo({
      db,
      payload,
      force: forceUndo,
      outputArgs: args,
    });
  }

  const conflictStrategy = args.conflict ?? (args.merge ? "merge" : undefined);
  const doc = await addDoc({
    db,
    payload,
    conflictStrategy,
    outputArgs: args,
  });

  // if addDoc has changed the id (e.g. because it relies on the modifiyTime), update lastDocRef again
  // TODO: if changing lastDoc to history may need to change this to overwrite first update
  if (doc._id !== payload._id) {
    await updateLastDocsRef(db, doc._id);
  }

  return doc;
}
