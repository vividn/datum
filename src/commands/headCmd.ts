import { tailArgs, tailCmd, TailCmdArgs } from "./tailCmd";
import { EitherDocument } from "../documentControl/DatumDocument";
import { outputArgs } from "../input/outputArgs";
import { dbArgs } from "../input/dbArgs";
import { ArgumentParser } from "argparse";
import { parseIfNeeded } from "../utils/parseIfNeeded";

export const headCmdArgs = new ArgumentParser({
  description: "show the first occurred/modified/created entries in the db",
  prog: "datum head",
  usage: "%(prog)s [field]",
  parents: [tailArgs, dbArgs, outputArgs],
});

export type HeadCmdArgs = TailCmdArgs;

export async function headCmd(
  args: HeadCmdArgs | string | string[],
  preparsed?: Partial<HeadCmdArgs>,
): Promise<EitherDocument[]> {
  args = parseIfNeeded(headCmdArgs, args, preparsed);
  return tailCmd({ ...args, head: true });
}
