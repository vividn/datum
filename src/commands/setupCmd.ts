import { Argv } from "yargs";
import { BaseDatumArgs } from "../input/baseYargs";
import { setupDatumViews } from "../views/setupDatumViews";
import { connectDb } from "../auth/connectDb";
import { Show } from "../output/output";

export const command = "setup";
export const desc = "setup the database for use with datum";

export type SetupCmdArgs = BaseDatumArgs & {
  projectDir?: string;
};

export function builder(yargs: Argv): Argv {
  return yargs.options({
    "project-dir": {
      describe:
        "where to look for additional DatumViews to setup in the database",
      type: "string",
    },
  });
}

export async function setupCmd(args: SetupCmdArgs): Promise<void> {
  const db = connectDb(args);
  const show: Show = args.showAll ? Show.All : args.show ?? Show.None;
  await setupDatumViews({ db, show: show, projectDir: args.projectDir });
}
