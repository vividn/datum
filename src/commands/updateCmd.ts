import { dataArgs, DataArgs, handleDataArgs } from "../input/dataArgs";
import {
  updateStrategies,
  UpdateStrategyNames,
} from "../documentControl/combineData";
import { DatumData, EitherDocument } from "../documentControl/DatumDocument";
import { connectDb } from "../auth/connectDb";
import { updateDoc } from "../documentControl/updateDoc";
import { QuickIdArgs, quickIdArgs } from "../input/quickIdArg";
import { flexiblePositional } from "../input/flexiblePositional";
import { updateLastDocsRef } from "../documentControl/lastDocs";
import { dbArgs } from "../input/dbArgs";
import { outputArgs } from "../input/outputArgs";
import { ArgumentParser } from "argparse";
import { MainDatumArgs } from "../input/mainArgs";
import { parseIfNeeded } from "../utils/parseIfNeeded";
import { quickId, _LAST_WITH_PROTECTION } from "../ids/quickId";
import { JsonType } from "../utils/utilityTypes";

export const updateArgs = new ArgumentParser({
  add_help: false,
  parents: [quickIdArgs, dataArgs, dbArgs, outputArgs],
});
updateArgs.add_argument("--strategy", "-X", {
  help: "which update strategy to use when modifying the doc",
  choices: Object.keys(updateStrategies),
});

export const updateCmdArgs = new ArgumentParser({
  description:
    "Update the data in an existing document." +
    " Defaults to 'preferNew' strategy for update command." +
    " Defaults to 'merge' for the merge command",
  prog: "dtm update/merge",
  usage: `%(prog)s <quickId> [data..]
  %(prog)s --strategy <strategy> <quickId> [data..]`,
  parents: [updateArgs],
});

export type UpdateCmdArgs = MainDatumArgs &
  DataArgs &
  QuickIdArgs & {
    strategy?: UpdateStrategyNames;
  };

export async function updateCmd(
  argsOrCli: UpdateCmdArgs | string | string[],
  preparsed?: Partial<UpdateCmdArgs>,
): Promise<EitherDocument[]> {
  const args = parseIfNeeded(updateCmdArgs, argsOrCli, preparsed);
  const db = connectDb(args);

  // process quickIds like the first required argument so that data changes can be specified beforehand in the command
  // for easier aliasing
  flexiblePositional(args, "quickId", `__quickId=${_LAST_WITH_PROTECTION}`);
  const { __quickId, ...payload }: DatumData<{ __quickId?: JsonType }> =
    handleDataArgs(args);

  const ids = await quickId(__quickId ?? _LAST_WITH_PROTECTION, args);

  // update now in case the updateDoc fails due to conflict
  await updateLastDocsRef(db, ids);

  const updateStrategy = args.strategy ?? "preferNew";

  const updatedDocs = await Promise.all(
    ids.map((id) =>
      updateDoc({
        db,
        id,
        payload,
        updateStrategy,
        outputArgs: args,
      }),
    ),
  );

  const newIds = updatedDocs.map((doc) => doc._id);
  if (newIds !== ids) {
    // TODO: if changing lastDocs to history may need to change this to overwrite first update
    await updateLastDocsRef(db, newIds);
  }

  return updatedDocs;
}
