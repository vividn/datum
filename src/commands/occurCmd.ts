import { handleTimeArgs, timeArgs, TimeArgs } from "../input/timeArgs";
import { addCmd, AddCmdArgs, newDocArgs } from "./addCmd";
import { EitherDocument } from "../documentControl/DatumDocument";
import { ArgumentParser } from "argparse";
import { parseIfNeeded } from "../utils/parseIfNeeded";
import { dbArgs } from "../input/dbArgs";
import { outputArgs } from "../input/outputArgs";
import { dataArgs } from "../input/dataArgs";
import { fieldArgs } from "../input/fieldArgs";

export const occurArgs = new ArgumentParser({
  add_help: false,
  parents: [fieldArgs, newDocArgs, timeArgs, dataArgs],
});
export const occurCmdArgs = new ArgumentParser({
  description: "record the occurence of something at a single time point",
  prog: "datum occur",
  usage: `%(prog)s <field> [data..]
  %(prog)s --fieldless [data..]
  %(prog)s <field> -K <reqKey1> ... -K <reqKeyN> -k <optKey1>[=defaultVal1] ... -k <optKeyN> <reqVal1> ... <reqValN> [optVal1] ... [optValN] [data..]
`,
  parents: [occurArgs, dbArgs, outputArgs],
});

export type OccurCmdArgs = AddCmdArgs & TimeArgs;

export async function occurCmd(
  args: OccurCmdArgs | string | string[],
  preparsed?: Partial<OccurCmdArgs>,
): Promise<EitherDocument> {
  args = parseIfNeeded(occurCmdArgs, args, preparsed);
  const { time: occurTime } = handleTimeArgs(args);
  if (occurTime !== undefined) {
    args.cmdData ??= {};
    args.cmdData.occurTime = occurTime;
  }

  return await addCmd(args);
}
