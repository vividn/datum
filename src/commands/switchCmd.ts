import { occurCmd, OccurCmdArgs } from "./occurCmd";
import { EitherDocument } from "../documentControl/DatumDocument";
import { flexiblePositional } from "../input/flexiblePositional";
import { durationArgs, DurationArgs } from "../input/durationArgs";

import { DatumState, normalizeState } from "../state/normalizeState";
import { ArgumentParser } from "argparse";
import { parseIfNeeded } from "../utils/parseIfNeeded";
import { dataArgs } from "../input/dataArgs";
import { timeArgs } from "../input/timeArgs";
import { fieldArgs } from "../input/fieldArgs";
import { dbArgs } from "../input/dbArgs";
import { outputArgs } from "../input/outputArgs";
import { newDocArgs } from "./addCmd";
import { inferType } from "../utils/inferType";

export const stateArgs = new ArgumentParser({
  add_help: false,
});
stateArgs.add_argument("state", {
  help: "the state to switch to, it defaults to true--equivalent to start",
  nargs: "?",
  default: "true",
});
stateArgs.add_argument("--last-state", "-L", {
  help: "manually specify the last state being transitioned out of",
  dest: "lastState",
});

export const switchArgs = new ArgumentParser({
  add_help: false,
  parents: [fieldArgs, stateArgs, durationArgs, newDocArgs, timeArgs, dataArgs],
});

export const switchCmdArgs = new ArgumentParser({
  description: "switch states of a given field",
  prog: "datum switch",
  usage: `%(prog)s <field> <state> [<duration OR .> [data..]]
  %(prog)s --moment <field> <state> [data..]
  %(prog)s <field> -k <reqKey> -k <optKey>=[defaultValue] ... <state> <duration OR .> <reqValue> [optValue] [data..]
`,
  parents: [switchArgs, dbArgs, outputArgs],
});

export type StateArgs = {
  state?: DatumState | string;
  lastState?: DatumState | string;
};
export type SwitchCmdArgs = OccurCmdArgs & DurationArgs & StateArgs;

export async function switchCmd(
  args: SwitchCmdArgs | string | string[],
  preparsed?: Partial<SwitchCmdArgs>,
): Promise<EitherDocument> {
  args = parseIfNeeded(switchCmdArgs, args, preparsed);
  flexiblePositional(
    args,
    "duration",
    "dur=",
    args.moment || args.omitTimestamp,
  );
  flexiblePositional(args, "state", "state=true");
  args.cmdData ??= {};
  if (args.moment) {
    args.cmdData.dur = null;
  }
  if (args.lastState) {
    const lastState =
      typeof args.lastState === "string"
        ? inferType(args.lastState)
        : args.lastState;
    if (lastState !== undefined) {
      args.cmdData.lastState = normalizeState(lastState);
    }
  }

  return await occurCmd(args);
}
