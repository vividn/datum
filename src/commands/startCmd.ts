import { Argv } from "yargs";
import { EitherDocument } from "../documentControl/DatumDocument";
import { switchCmd } from "./switchCmd";
import { occurArgs, OccurCmdArgs } from "./occurCmd";
import { DurationArgs, durationArgs } from "../input/durationArgs";

export const command = "start <field> [duration] [data..]";
export const desc = "add a start document";

export function durOccurArgs(yargs: Argv): Argv {
  return durationArgs(occurArgs(yargs));
}
export function builder(yargs: Argv): Argv {
  return durOccurArgs(yargs);
}

export type StartCmdArgs = OccurCmdArgs & DurationArgs;

export async function startCmd(args: StartCmdArgs): Promise<EitherDocument> {
  return await switchCmd({ ...args, state: true });
}
