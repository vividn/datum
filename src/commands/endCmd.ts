import { Argv } from "yargs";
import { EitherDocument } from "../documentControl/DatumDocument";
import { durOccurArgs, StartCmdArgs } from "./startCmd";
import { switchCmd } from "./switchCmd";
import set from "lodash.set";

export const command = "end <field> [data..]";
export const desc = "add an end document";

export function builder(yargs: Argv): Argv {
  return durOccurArgs(yargs);
}

export type EndCmdArgs = StartCmdArgs;

export async function endCmd(args: StartCmdArgs): Promise<EitherDocument> {
  args.cmdData ??= {};
  set(args.cmdData, "state.id", false);
  return await switchCmd({ ...args });
}
