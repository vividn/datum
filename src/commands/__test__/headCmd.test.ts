import {
  coloredChalk,
  deterministicHumanIds,
  mockedLogLifecycle,
  mockSpecs,
  setNow,
  testDbLifecycle,
} from "../../__test__/test-utils";
import { endCmd } from "../endCmd";
import { startCmd } from "../startCmd";
import { switchCmd } from "../switchCmd";
import { occurCmd } from "../occurCmd";
import { Show } from "../../input/outputArgs";
import { setupCmd } from "../setupCmd";
import { addCmd } from "../addCmd";
import { updateCmd } from "../updateCmd";
import { DateTime, Settings } from "luxon";
import { headCmd } from "../headCmd";
import { generateSampleMorning } from "../../__test__/generateSampleMorning";
import { datumTimeToLuxon } from "../../time/datumTime";
import { DatumDocument } from "../../documentControl/DatumDocument";

const yesterday = "2023-10-31";
const today = "2023-11-01";
const tomorrow = "2023-11-02";

describe("headCmd", () => {
  const { mockedLog } = mockedLogLifecycle();
  deterministicHumanIds();
  coloredChalk();
  const dbName = "head_cmd_test";
  testDbLifecycle(dbName);
  mockSpecs();

  beforeEach(async () => {
    await setupCmd("");
    setNow(`20:00 ${today}`);
  });

  it.todo("calls tailCmd with head: true and other arguments passed through");

  it("displays by default the first 10 occurrences in the database", async () => {
    await generateSampleMorning(today);
    const docs = await headCmd("", { show: Show.Standard });
    expect(docs.length).toBe(10);
    expect(docs[0]._id).toMatchInlineSnapshot(
      `"sleep:2023-10-31T23:20:00.000Z"`,
    );
    expect(docs.at(-1)?._id).toMatchInlineSnapshot(
      `"stretch:2023-11-01T09:37:00.000Z"`,
    );
    expect(mockedLog.mock.calls).toMatchSnapshot();
  });

  it("will display all if there are less than 10 occurrences in db", async () => {
    setNow(`8am ${today}`);
    await occurCmd("caffeine amount=100");
    setNow(`10am`);
    await switchCmd("project household");
    setNow(`10:30`);
    await endCmd("project");

    const docs = await headCmd("", { show: Show.Standard });
    expect(docs.length).toBe(3);
    expect(mockedLog.mock.calls).toMatchSnapshot();
  });

  it("can display the first n occurrences", async () => {
    await generateSampleMorning(today);
    const docs = await headCmd("-n 5");
    expect(docs.length).toBe(5);
  });

  it("will display all if there are less than n occurrences in db", async () => {
    setNow(`8am ${today}`);
    await occurCmd("caffeine amount=100");
    setNow(`10am`);
    await switchCmd("project household");
    setNow(`10:30`);
    await endCmd("project");

    const docs = await headCmd("-n 5");
    expect(docs.length).toBe(3);
  });

  it("displays first occurrences of a specific field", async () => {
    await generateSampleMorning(today);
    const docs = await headCmd("stretch");
    expect(docs.length).toBe(4);
    await generateSampleMorning(yesterday);
    const docs2 = (await headCmd("stretch")) as DatumDocument[];
    expect(docs2.length).toBe(8);
    expect(docs2.map((doc) => doc.data.field as string)).toEqual(
      Array(8).fill("stretch"),
    );
  });

  describe("first metrics", () => {
    let occur1Id: string, occur2Id: string, occur3Id: string, addId: string;
    beforeEach(async () => {
      setNow("19:00");
      ({ _id: occur1Id } = await occurCmd("alcohol"));
      setNow("20:00");
      ({ _id: addId } = await addCmd(
        "person name='john doe' age=35 --id %name",
      ));
      setNow("21:00");
      ({ _id: occur3Id } = await endCmd("socialize"));
      setNow("+1");
      ({ _id: occur2Id } = await startCmd("socialize -t 19:30"));

      // modify the doc without an occur time
      setNow("22:00");
      await updateCmd(`"${addId}" age=36`);
    });

    it("defaults to a hybrid display of occurTime and createTime", async () => {
      const docs = await headCmd("", { show: Show.Standard });
      expect(docs.length).toBe(4);
      expect(docs.map((doc) => doc._id)).toEqual([
        occur1Id,
        occur2Id,
        addId,
        occur3Id,
      ]);
      expect(mockedLog.mock.calls).toMatchSnapshot();

      const explicitHybridDocs = await headCmd("-m hybrid");
      expect(explicitHybridDocs).toEqual(docs);
    });

    it("can just display occurTime head", async () => {
      const docs = await headCmd("-m occur");
      expect(docs.length).toBe(3);
      expect(docs.map((doc) => doc._id)).toEqual([
        occur1Id,
        occur2Id,
        occur3Id,
      ]);
      expect(mockedLog.mock.calls).toMatchSnapshot();
    });

    it("can display modifyTime head", async () => {
      const docs = await headCmd("-m modify");
      expect(docs.length).toBe(4);
      expect(docs.map((doc) => doc._id)).toEqual([
        occur1Id,
        occur3Id,
        occur2Id,
        addId,
      ]);
      expect(mockedLog.mock.calls).toMatchSnapshot();
    });

    it("can display createTime head", async () => {
      const docs = await headCmd("-m create");
      expect(docs.length).toBe(4);
      expect(docs.map((doc) => doc._id)).toEqual([
        occur1Id,
        addId,
        occur3Id,
        occur2Id,
      ]);
    });
  });

  it("can display a head from a certain moment in time", async () => {
    await generateSampleMorning(today);
    const docs1 = await headCmd("-t 10:15");
    expect(docs1.length).toBe(8);
    expect(docs1.at(-1)?._id).toMatchInlineSnapshot(
      `"project:2023-11-01T11:45:00.000Z"`,
    );

    const docs2 = await headCmd(`-d today -t 11`);
    expect(docs2.length).toBe(4);

    const docs3 = await headCmd("-d +1 -t 7");
    expect(docs3.length).toBe(0);
  });

  it("displays all occurrences on a day if date is given without time", async () => {
    await generateSampleMorning(today);
    await generateSampleMorning(yesterday);
    const docs = (await headCmd(`-d ${today}`)) as DatumDocument[];
    const yesterdayDocs = (await headCmd("-y")) as DatumDocument[];

    expect(docs.length).toBeGreaterThan(10);
    expect(
      docs
        .map((doc) => doc.data.occurTime?.utc)
        .every(
          (occurTime) =>
            DateTime.fromISO(occurTime ?? "").toISODate() === today,
        ),
    ).toBe(true);
    expect(
      yesterdayDocs
        .map((doc) => doc.data.occurTime?.utc)
        .every(
          (occurTime) =>
            DateTime.fromISO(occurTime ?? "").toISODate() === yesterday,
        ),
    ).toEqual(true);
  });

  it("displays occurrences on a given day properly even when user is in a different timezone", async () => {
    // use New Zealand so that early events in the sample morning are previous day in UTC
    Settings.defaultZone = "Pacific/Auckland";
    setNow(`${today} 20:00`);
    await generateSampleMorning(today);
    await generateSampleMorning(tomorrow);
    const docs = (await headCmd(`-d today`)) as DatumDocument[];
    expect(docs.length).toBeGreaterThan(10);

    // when mapping just from utc, not all should be on today
    const utcDates = docs.map((doc) => {
      const utc = doc.data.occurTime?.utc;
      if (!utc) {
        throw Error("bad occurTime");
      }
      return DateTime.fromISO(utc).toUTC().toISODate();
    });
    expect(utcDates.every((date) => date === today)).toBe(false);

    const localDates = docs.map((doc) =>
      datumTimeToLuxon(doc.data.occurTime)?.toISODate(),
    );
    expect(localDates.every((date) => date === today)).toBe(true);

    Settings.defaultZone = "system";
  });

  it("when requesting the entire date, have full day occurrences be at the top", async () => {
    Settings.defaultZone = "Pacific/Auckland";
    setNow(`${today} 20:00`);
    await generateSampleMorning(today);
    await generateSampleMorning(tomorrow);
    const lengthWithoutFullDayDocs = (await headCmd("-d +1")).length;

    const fullDayDoc1 = await occurCmd("field -d +1");
    const fullDayDoc2 = await occurCmd(`otherField -d ${tomorrow}`);
    const docs = await headCmd("-d +1");
    expect(docs.length).toEqual(lengthWithoutFullDayDocs + 2);

    expect(docs.slice(0, 2)).toEqual([fullDayDoc1, fullDayDoc2]);
    Settings.defaultZone = "system";
  });

  it("displays only first n occurrences on a day if date and -n is given", async () => {
    Settings.defaultZone = "Pacific/Auckland";
    setNow(`${today} 20:00`);
    await generateSampleMorning(today);
    await generateSampleMorning(tomorrow);

    const fullDayDoc1 = await occurCmd("field -d +1");
    const fullDayDoc2 = await occurCmd("otherField -d +1");
    const allDocs = await headCmd("-d +1");
    const limitedDocs = await headCmd("-d +1 -n 5");

    expect(limitedDocs.length).toEqual(5);
    expect(limitedDocs[0]).toEqual(fullDayDoc1);
    expect(limitedDocs[1]).toEqual(fullDayDoc2);
    expect(limitedDocs).toEqual(allDocs.slice(0, 5));
    Settings.defaultZone = "system";
  });

  it("can display a custom format for the head commands", async () => {
    await generateSampleMorning(today);
    await generateSampleMorning(tomorrow);

    await occurCmd("", { show: Show.None });
    await headCmd(" --format '::%field%::'");
    expect(mockedLog.mock.calls).toMatchInlineSnapshot(`
      [
        [
          "::sleep::
      ::sleep::
      ::project::
      ::text::
      ::project::
      ::text::
      ::environment::
      ::stretch::
      ::run::
      ::stretch::",
        ],
      ]
    `);
  });

  it("does not display anything when show is None", async () => {
    await generateSampleMorning(today);
    await headCmd("", { show: Show.None });
    expect(mockedLog).not.toHaveBeenCalled();
  });
});
