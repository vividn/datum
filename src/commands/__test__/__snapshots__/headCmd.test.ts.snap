// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`headCmd displays by default the first 10 occurrences in the database 3`] = `
[
  [
    "          time  field        state           hid  
-1d 23:20:00[90m+0[39m  [7msleep[27m        [1mstart[22m           en61v
    08:30:00[90m+0[39m  [7msleep[27m        [1mend[22m             elfuv
    08:40:00[90m+0[39m  [7mproject[27m      german          d7j7m
    08:40:00[90m+0[39m  [7mtext[27m         fiction_book    w1fla
    09:10:00[90m+0[39m  [7mproject[27m      [1mend[22m             mbukz
    09:10:00[90m+0[39m  [7mtext[27m         [1mend[22m             puzt1
    09:30:00[90m+0[39m  [7menvironment[27m  outside         wqynb
    09:30:00[90m+0[39m  [7mstretch[27m      [1mstart[22m           7qaqo
    09:37:00[90m+0[39m  [7mrun[27m          [1mstart[22m           9bcqg
    09:37:00[90m+0[39m  [7mstretch[27m      [1mend[22m             6m3y5
",
  ],
]
`;

exports[`headCmd first metrics can display modifyTime head 1`] = `
[
  [
    "       time  field      state    hid  
[90mm[39m19:00:00[90m+0[39m  [7malcohol[27m             elfuv
[90mm[39m21:00:00[90m+0[39m  [7msocialize[27m  [1mend[22m      d7j7m
[90mm[39m21:01:00[90m+0[39m  [7msocialize[27m  [1mstart[22m    w1fla
[90mm[39m22:00:00[90m+0[39m  [7mperson[27m              en61v
",
  ],
]
`;

exports[`headCmd first metrics can just display occurTime head 1`] = `
[
  [
    "      time  field      state    hid  
19:00:00[90m+0[39m  [7malcohol[27m             elfuv
19:30:00[90m+0[39m  [7msocialize[27m  [1mstart[22m    w1fla
21:00:00[90m+0[39m  [7msocialize[27m  [1mend[22m      d7j7m
",
  ],
]
`;

exports[`headCmd first metrics defaults to a hybrid display of occurTime and createTime 1`] = `
[
  [
    "       time  field      state    hid  
 19:00:00[90m+0[39m  [7malcohol[27m             elfuv
 19:30:00[90m+0[39m  [7msocialize[27m  [1mstart[22m    w1fla
[90mc[39m20:00:00[90m+0[39m  [7mperson[27m              en61v
 21:00:00[90m+0[39m  [7msocialize[27m  [1mend[22m      d7j7m
",
  ],
]
`;

exports[`headCmd will display all if there are less than 10 occurrences in db 1`] = `
[
  [
    "      time  field     state        hid  
08:00:00[90m+0[39m  [7mcaffeine[27m               elfuv
10:00:00[90m+0[39m  [7mproject[27m   household    en61v
10:30:00[90m+0[39m  [7mproject[27m   [1mend[22m          d7j7m
",
  ],
]
`;
